File: linkedlist/singly/SinglyLinkedList.java
/**
 * @author Ayomide Oyekanmi
 */
class SinglyLinkedList<E> {

    /** 
     * This represents the template for the node object that is used in the linkedlist
     */
    class Node<E> {
        E element;
        Node<E> next;

        public Node(E element){
            this.element = element;
            next = null;
        }
    }

    private Node<E> head;
    private Node<E> tail;
    private int size;

    public SinglyLinkedList(){ 
        head = null;
        tail = null;
        size = 0;
    }

    /**
     * Add element to front of the list, so that it becomes new head node
     * 
     * @param element element that the new node created wraps around
     */
    public void addFirst(E element){
        // Create new node to wrap element
        Node<E> newHead = new Node(element);
        // Set the new node's pointer to the current head
        newHead.next = head;
        // Set new node as head
        head = newHead;

        // If tail hasn't be set, tail is equal to head
        if(tail == null){
            tail = head;
        }

        // Increment size
        size++;
    }

    /**
     * Add element to back of the list, so that it becomes new tail node
     * 
     * @param element element that the new node created wraps around
     */
    public void addLast(E element){
        // Starting at head node, traverse the linkedlist until tail node is reached
        Node<E> node = head;
        while(node.next != null){
           node = node.next; 
        }

        // Set the current tail node's pointer to the new node that wraps around element
        node.next = new Node(element);
        // Set new node as new tail node
        tail = node.next;

        // Increment size
        size++;
    }

    /**
     * At element at any position in the list
     * 
     * @param n represents position to insert element into.
     * @param element represents element to be inserted into position n
     * @throws IllegalArgumentException if the position given does not exist for list
     */
    public void addAtPosition(int n, E element){
        // Checks if position exists
        if (n < 1 || n > size){
            IllegalArgumentException("No such position");
        }

        // Checks if element to be added will be the new head
        if (n == 1){
            addFirst(element);
            return;
        }

        int index = 1;
        Node<E> prev = null;
        Node<E> node = head;
        Node<E> newNode = new Node(element);

        // Traverse till we get to position
        // PS: node != null check is not necessary because we already check if given index is valid
        // using if (n > size) 
        while(index != n){
            prev = node;
            node = node.next;
        }

        // Update pointers to reference new node
        prev.next =  newNode;
        newNode.next = node;

        // Increment size
        size++;
    }

    /**
     * Removes head of the list and returns it
     * 
     * @return returns removed element
     * @throws NoSuchElementException if list has no head i.e. the list is empty
     */
    public E removeFirst(){
        // Checks if list is empty
        if(head == null){
            throw new NoSuchElementException();
        }

        // Checks if head element is the only element in the list
        if(head.next == null){
            head = null;
            tail = null;
            return;
        }

        // Removes reference to head element
        Node<E> nodeToRemove = head;
        head = head.next;

        // Decrements size
        size--;

        return nodeToRemove;
    }

    /**
     * Removes tail of the list and returns it
     * 
     * @return returns removed element
     * @throws NoSuchElementException if list has no tail i.e. the list is empty
     */
    public E removeLast(){
        // Checks if list is empty
        if (tail == null){
            throw new NoSuchElementException();
        }

        // Checks if tail is the only element in the list
        // If head.next is null, it means head == tail :)
        if(head.next == null){
            head = null;
            tail = null;
            return;
        }

        // Traverses to the end of the list
        Node<E> prev = null;
        Node<E> node = head;
        while(node.next != null){
            prev = node;
            node = node.next;
        }

        // Removes reference to current tail
        prev.next = null;
        // Set the new tail
        tail = prev;

        // Decrements size
        size--;

        return node;
    }

    /**
     * Removes element at position n in the list and returns it
     * 
     * @param n position to remove element from
     * @return returns removed element
     * @throws IllegalArgumentException if the position given does not exist for list
     */
    public E removeAtPosition(int n){
        // Checks if position exists
        if (n < 1 || n > size){
            IllegalArgumentException("No such position");
        }

        // Checks if element to be removed is head
        if(n == 0){
            removeFirst();
            return;
        }

        // Checks if element to be removed is tail
        if(n == size){
            removeLast();
            return;
        }

        
        int index = 1;
        Node<E> prev = null;
        Node<E> node = head;

        // Traverse till we get to position
        // PS: node != null check is not necessary because we already check if given index is valid
        // using if (n > size) 
        while(index != n){
            prev = node;
            node = node.next;
        }

        // Remove reference to node to be deleted
        prev.next = node.next;

        // Decrement size
        size--;

        return node;
    }
}

/**
 * The SinglyLinkedList object will be instantiated and called as such:
 * SinglyLinkedList<String> obj = new SinglyLinkedList<>();
 * obj.addFirst('Ayo'); # Ayo
 * obj.addFirst('Yuwa'); # Yuwa -> Ayo
 * obj.addLast('Jess'); # Yuwa -> Ayo -> Jess
 * obj.addLast('Shay'); # Yuwa -> Ayo -> Jess -> Shay
 * obj.addAtPosition(3, 'Vee') # Yuwa -> Ayo -> Vee -> Jess -> Shay
 * obj.removeFirst() # Ayo -> Vee -> Jess -> Shay
 * obj.removeLast() # Ayo -> Vee -> Jess
 * obj.removeAtPosition(2) # Ayo -> Jess
 */

File: linkedlist/singly/python/SinglyLinkedList.py
"""
Author: Abdulazeez Abdulazeez Adeshina
"""


class Node:
    """
    This represents the template for the node object that is used for the linkedlist.
    """

    def __init__(self, element):
        """
        :param element: This represents the element contained in the Node.
        :type element: This element can be of any type ranging from strings, numbers, arrays etc.
        """
        self.element = element
        self.next = None


class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size: int = 0


    def __len__(self) -> int:
        """
        :return: Length of the linkedlist
        :rtype: int
        """
        return self.size

    def addFirst(self, element):
        """
        The `addFirst` method adds a new node to the front of the list making it the head of the linkedlist.

        :param element: the value held in the Node.
        """
        # Create a new node to wrap the element
        newHead = Node(element)
        # Set the new node's next pointer to the current head
        newHead.next = self.head
        # Set the new node as the head
        self.head = newHead

        # Check if the tail node is absent.
        if self.tail == None:
            # If the tail node is absent, assign the head node to the tail node.
            self.tail = self.head

        # Increment linked list size.
        self.size += 1

    def addLast(self, element):
        """
        The `addLast` method adds a new node to the end of the list.

        :param element: the value held in the Node.
        """

        # Starting at the head node, travserse the linkedlist until we get to the tail node
        node = self.head

        while node.next != None:
            node = node.next

        # Set the current tail node's pointer to the new tail node

        node.next = Node(element)
        # Set the new tail node

        self.tail = node.next

        # Increment size
        self.size += 1

    def addAtPosition(self, position, element):
        """
        The `addAtPosition` method adds a new node to the specified position in the linkedlist if the position is valid.

        :param position: position of the new element to be added.
        :param element: the value held in the Node.
        :raise
        """

        # Check if position is valid
        if position < 1 or position > self.size:
            raise IndexError("Position is invalid")

        # Check of element to be added will be the new head
        if position == 1:
            self.addFirst(element)
            return

        index = 1
        prev = None
        current = self.head
        newNode = Node(element)

        # Traverse the linkedlist till we get to the position
        # using if (position > self.size)
        while index != position:
            index += 1
            prev = current
            current = current.next

        prev.next = newNode
        newNode.next = current

        self.size += 1

    def removeFirst(self):
        """
        The `removeFirst` method removes the head of the linkedlist and returns it.

        :returns removed element
        """

        # Check if the list is empty
        if self.head is None:
            return "List is Empty!"

        # Check if the head element is the only element in the linkedlist
        if self.head.next is None:
            self.head = self.tail = None
            self.size = 0
            return

        # Removes reference to head element
        nodeToRemove = self.head
        self.head = self.head.next

        # Decrement size
        self.size -= 1

        return nodeToRemove.element

    def removeLast(self):
        """
        The `removeLast` method removes the tail of the linkedlist.

        :returns the removed element
        """

        # Check if linkedlist is empty
        if self.tail is None:
            return "No tail present in linkedlist!"

        # Check if the tail is the only element in the list
        # If `self.head.next` is None, it means `self.head` == `self.tail` :)

        if self.head.next is None:
            self.head = None
            self.tail = None
            self.size = 0
            return

        prev = None
        current = self.head

        # Traverse to the end of the list
        while current.next is not None:
            prev = current
            current = current.next

        # Remove reference to current tail
        prev.next = None

        # Set the new tail
        self.tail = prev

        # Decrement size
        self.size -= 1

        return current.element

    def removeAtPosition(self, position):
        """
        The `removeAtPosition` method removes a node at the specified postion.

        :param position: position to remove element from
        :returns the removed element
        """

        # Check if position is valid
        if position < 1 or position > self.size:
            raise IndexError("Invalid position!")

        # Check if element to be removed is the head
        if position == 0:
            self.removeFirst()
            return

        # Check if element to be removed is the tail
        if position == self.size:
            self.removeLast()
            return

        index = 1
        prev = None
        current = self.head

        # Traverse till we get to position
        while index != position:
            index += 1
            prev = current
            current = current.next

        # Remove reference to node to be deleted
        prev.next = current.next

        # Decrement size
        self.size -= 1

        return current.element

if __name__ == '__main__':
    x = SinglyLinkedList()
    x.addFirst(1) # 1
    x.addLast(2) # 1 -> 2
    x.addFirst(3) # 3 -> 1 -> 2
    x.addAtPosition(2, 10) # 3 -> 10 -> 1 -> 2
    x.removeAtPosition(2) # 3 -> 1 -> 2
    print(len(x)) # 3
